keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -0,0 +1,730 @@
++ _BucketClass (BC)_ - a cluster-scoped custom resource containing fields defining the provisioner and an immutable parameter set for creating new buckets.
++ _BucketRequest (BR)_ - a user-namespaced custom resource representing a request for a new backend bucket, or access to an existing bucket.
++ _COSI_ - Container _Object_ Store Interface, modeled after CSI.
++ _cosi-node-adapter_ - a pod per node which receives Kubelet gRPC _NodeStageVolume_ and _NodeUnstageVolume_ requests, ensures the target bucket has been provisioned, and notifies the kubelet when the pod can be run.
++ _driver_ - a container (usually paired with a sidecar container) that is responsible for communicating with the underlying object store to create, delete, grant access to, and revoke access from buckets. Drivers talk gRPC and need no knowledge of Kubernetes. Drivers are typically written by storage vendors, and should not be given any access outside of their namespace.
++ _driverless_ - a use-case where no driver exists to support the backend object store. Some COSI automation may still be provided, but backend operations such as creating a bucket or minting credentials will not occur.
++ _greenfield bucket_ - a new backend bucket created by COSI.
++ _green-to-brownfield_ - a use-case where COSI creates a new bucket on behalf of a user, and then supports ways for other users in the cluster to share this bucket.
++ _provisioner_ - a generic term meant to describe the combination of a sidecar and driver. "Provisioning" a bucket can mean creating a new bucket or granting access to an existing bucket.
++ _sidecar_ - a Kubernetes-aware container (usually paired with a driver) that is responsible for fullfilling COSI requests with the driver via gRPC calls. The sidecar's access can be restricted to the namespace where it resides, which is expected to be the same namespace as the provisioner. The COSI sidecar is provided by the Kubernetes community.
++ _static provisioning_ - custom resource creation is done by the admin rather than via COSI automation. This may also include _driverless_ but they are independent.
+
+# Proposal
+
+## APIs
+
+### Storage APIs
+
+#### BucketRequest
+
+A user facing, namespaced custom resource requesting provisioning of a new bucket, or requesting access to an existing bucket. A `BucketRequest` (BR) lives in the app's namespace.  In addition to a `BucketRequest`, a [BucketAccessRequest](#bucketaccessrequest) is necessary in order to grant credentials to access the bucket. BRs are required for both greenfield and brownfield uses.
+
+There is a 1:1 mapping of a `BucketRequest` and the cluster scoped `Bucket` instance, meaning there is a single `Bucket` instance for every BR.
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: BucketRequest
+metadata:
+  name:
+  namespace:
+  labels:
+    - cosi.io/provisioner: [1]
+  finalizers:
+    - cosi.io/finalizer [2]
+spec:
+  protocol:
```

Comment:
We should change this to a list of protocols: to support asking for multiple protocols. Effectively this is a way for an app to say I understand this set of protocols.

Why? Imagine you have an app that is deployed across multiple cloud providers, and bundles multiple SDKs to speak the native object protocol on each environment. The app dev should be able to define a single `BucketRequest` that will work across all those environments.

Reply:
- use multiple BCs (single protocol per BC) instead of list of protocols
- remove BR.protocol
- no impact on app portability
- single protocol passed to driver in case driver supports multiple protocols
Earlier drafts had a list of protocols in the BC and a single protocol in the BR. During reviews we concluded that if a vendor supports multiple protocols the easiest solution is to create 1 BC per protocol. We believe this approach is more intuitive and less complex. `BC.protocol` is passed to the driver which can reject the request if the protocol is not supported.

Along the lines of intuitive and less complex, we are considering eliminating protocol from the BR. The reason why is that the BC reference in the BR is sufficient to know the protocol. The only value today in `BR.protocol` is that COSI can validate the BC's protocol matches the BR's. If validation fails then the BR is retried as is typical in k8s. An issue with this is the BC is immutable so the admin would need to recreate the BC with the correct protocol. Or, the BR needs to be deleted and corrected to use the right protocol. Both of these scenarios are unnecessary if `BR.prtocol` is removed.

Thoughts?

Reply:
I think the use case I described above is an important one:

> Imagine you have an app that is deployed across multiple cloud providers, and bundles multiple SDKs to speak the native object protocol on each environment. The app dev should be able to define a single BucketRequest that will work across all those environments.

Having to create a new BR for each platform breaks the portability we are trying to achieve.

Reply:
@saad-ali Are you describing a single app that speaks multiple protocols, ie .accessing >1 bucket and using a different protocol among buckets?  

But I think regardless of the answer above, the app can reference 2+ BARs in thei csi-driver portion of the pod spec. Each BAR references a BR. Each BR references a BC and the BC defines a single protocol (according to the current kep). I don't see a portability issue since we assume that the admin sets up the correct BCs and provisioners on the target cluster.